Mathlib/Order/BooleanGenerators.lean

```diff
@@ -0,0 +1,189 @@
+(see `IsCompactlyGenerated.BooleanGenerators.booleanAlgebra_of_sSup_eq_top`).
+-/
+structure BooleanGenerators (S : Set α) : Prop where
+  /-- The elements in a collection of boolean generators are all atoms. -/
+  isAtom : ∀ I ∈ S, IsAtom I
+  /-- The elements in a collection of boolean generators satisfy an atomicity condition:
+  any compact element below the supremum of a finite subset `s` of generators
+  is equal to the supremum of a subset of `s`. -/
+  finitelyAtomistic : ∀ (s : Finset α) (a : α),
+      ↑s ⊆ S → IsCompactElement a → a ≤ s.sup id → ∃ t ⊆ s, a = t.sup id
+
+namespace BooleanGenerators
+
+variable {S : Set α} (hS : BooleanGenerators S)
+
+lemma mono {T : Set α} (hTS : T ⊆ S) : BooleanGenerators T where
+  isAtom I hI := hS.isAtom I (hTS hI)
+  finitelyAtomistic := fun s a hs ↦ hS.finitelyAtomistic s a (le_trans hs hTS)
+
+lemma atomistic (a : α) (ha : a ≤ sSup S) : ∃ T ⊆ S, a = sSup T := by
+  obtain ⟨C, hC, rfl⟩ := IsCompactlyGenerated.exists_sSup_eq a
+  have aux : ∀ b : α, IsCompactElement b → b ≤ sSup S → ∃ T ⊆ S, b = sSup T := by
+    intro b hb hbS
+    obtain ⟨s, hs₁, hs₂⟩ := hb S hbS
+    obtain ⟨t, ht, rfl⟩ := hS.finitelyAtomistic s b hs₁ hb hs₂
+    refine ⟨t, ?_, Finset.sup_id_eq_sSup t⟩
+    refine Set.Subset.trans ?_ hs₁
+    simpa only [Finset.coe_subset] using ht
+  choose T hT₁ hT₂ using aux
+  use sSup {T c h₁ h₂ | (c ∈ C) (h₁ : IsCompactElement c) (h₂ : c ≤ sSup S)}
+  constructor
+  · apply _root_.sSup_le
+    rintro _ ⟨c, -, h₁, h₂, rfl⟩
+    apply hT₁
+  · apply le_antisymm
+    · apply _root_.sSup_le
+      intro c hc
+      rw [hT₂ c (hC _ hc) ((le_sSup hc).trans ha)]
+      apply sSup_le_sSup
+      apply _root_.le_sSup
+      use c, hc, hC _ hc, (le_sSup hc).trans ha
+    · simp only [Set.sSup_eq_sUnion, sSup_le_iff, Set.mem_sUnion, Set.mem_setOf_eq,
+        forall_exists_index, and_imp]
+      rintro a T b hbC hb hbS rfl haT
+      apply (le_sSup haT).trans
+      rw [← hT₂]
+      exact le_sSup hbC
+
+lemma isAtomistic_of_sSup_eq_top (h : sSup S = ⊤) : IsAtomistic α := by
+  refine ⟨fun a ↦ ?_⟩
+  obtain ⟨s, hs, hs'⟩ := hS.atomistic a (h ▸ le_top)
+  exact ⟨s, hs', fun I hI ↦ hS.isAtom I (hs hI)⟩
+
+lemma mem_of_isAtom_of_le_sSup_atoms (a : α) (ha : IsAtom a) (haS : a ≤ sSup S) :
+    a ∈ S := by
+  obtain ⟨T, hT, rfl⟩ := hS.atomistic a haS
+  obtain rfl | ⟨a, haT⟩ := T.eq_empty_or_nonempty
+  · simp only [sSup_empty] at ha
+    exact (ha.1 rfl).elim
+  suffices sSup T = a from this ▸ hT haT
+  have : a ≤ sSup T := le_sSup haT
+  rwa [ha.le_iff_eq, eq_comm] at this
+  exact (hS.isAtom a (hT haT)).1
+
+lemma sSup_le_sSup_iff_of_atoms (X Y : Set α) (hX : X ⊆ S) (hY : Y ⊆ S) :
+    sSup X ≤ sSup Y ↔ X ⊆ Y := by
+  refine ⟨?_, sSup_le_sSup⟩
+  intro h a ha
+  apply (hS.mono hY).mem_of_isAtom_of_le_sSup_atoms _ _ ((le_sSup ha).trans h)
+  exact (hS.mono hX).isAtom a ha
+
+lemma complementedLattice_of_sSup_eq_top (h : sSup S = ⊤) : ComplementedLattice α := by
+  constructor
```

Comment:
I can't help wondering if it might be easier to establish modularity instead. So the proof would start:
```suggestion
  suffices IsModularLattice α by
    have _i := isAtomistic_of_sSup_eq_top hS h
    apply complementedLattice_of_isAtomistic
```
More generally I'm not sure what order to establish properties would minimize the effort here (e.g., could we establish complementarity first and then get atomisticness from `complementedLattice_iff_isAtomistic` etc.).

However what you have is clearly fine.